src/S3/S3Transfer/MultipartDownloader.php

```diff
@@ -0,0 +1,427 @@
+<?php
+namespace Aws\S3\S3Transfer;
+
+use Aws\CommandInterface;
+use Aws\ResultInterface;
+use Aws\S3\S3ClientInterface;
+use Aws\S3\S3Transfer\Exceptions\S3TransferException;
+use Aws\S3\S3Transfer\Models\DownloadResult;
+use Aws\S3\S3Transfer\Models\S3TransferManagerConfig;
+use Aws\S3\S3Transfer\Progress\TransferListener;
+use Aws\S3\S3Transfer\Progress\TransferListenerNotifier;
+use Aws\S3\S3Transfer\Progress\TransferProgressSnapshot;
+use Aws\S3\S3Transfer\Utils\DownloadHandler;
+use GuzzleHttp\Promise\Coroutine;
+use GuzzleHttp\Promise\Create;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Promise\PromisorInterface;
+
+abstract class MultipartDownloader implements PromisorInterface
+{
+    public const GET_OBJECT_COMMAND = "GetObject";
+    public const PART_GET_MULTIPART_DOWNLOADER = "part";
+    public const RANGED_GET_MULTIPART_DOWNLOADER = "ranged";
+    private const OBJECT_SIZE_REGEX = "/\/(\d+)$/";
+    
+    /** @var array */
+    protected readonly array $getObjectRequestArgs;
+
+    /** @var array */
+    protected readonly array $config;
+
+    /** @var DownloadHandler */
+    private DownloadHandler $downloadHandler;
+
+    /** @var int */
+    protected int $currentPartNo;
+
+    /** @var int */
+    protected int $objectPartsCount;
+
+    /** @var int */
+    protected int $objectSizeInBytes;
+
+    /** @var string|null */
+    protected ?string $eTag;
+
+    /** @var TransferListenerNotifier | null */
+    private readonly ?TransferListenerNotifier $listenerNotifier;
+
+    /** Tracking Members */
+    private ?TransferProgressSnapshot $currentSnapshot;
+
+    /**
+     * @param S3ClientInterface $s3Client
+     * @param array $getObjectRequestArgs
+     * @param array $config
+     * @param DownloadHandler $downloadHandler
+     * @param int $currentPartNo
+     * @param int $objectPartsCount
+     * @param int $objectSizeInBytes
+     * @param string|null $eTag
+     * @param TransferProgressSnapshot|null $currentSnapshot
+     * @param TransferListenerNotifier|null $listenerNotifier
+     */
+    public function __construct(
+        protected readonly S3ClientInterface $s3Client,
+        array $getObjectRequestArgs,
+        array $config,
+        DownloadHandler $downloadHandler,
```

Comment:
should a default be instantiated if a `DownloadHandler` isn't provided? I feel like this is a little granular for most folks to try to reason with

Reply:
That makes sense. I have marked this parameter nullable and when found null internally it will be defaulted to StreamDownloadHandler.